pkg/collector/prometheus.go

```diff
@@ -0,0 +1,583 @@
+				c.nodeDesc.nodePackageMiliJoulesTotal,
+				prometheus.CounterValue,
+				float64(val.Pkg),
+				nodeName, strconv.Itoa(pkgID), coreEnergy, dramEnergy, uncoreEnergy,
+			) // deprecated metric
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		nodeEnergyStatusLabelValues := []string{nodeName, cpuArch}
+		for _, label := range nodeEnergyStatLabels[2:] {
+			val := uint64(nodeEnergy.Usage[label])
+			valStr := strconv.FormatUint(val, 10)
+			nodeEnergyStatusLabelValues = append(nodeEnergyStatusLabelValues, valStr)
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.nodeDesc.nodeEnergyStat,
+			prometheus.CounterValue,
+			(float64(nodeEnergy.SensorEnergy.Curr())/miliJouleToJoule)/SamplePeriodSec,
+			nodeEnergyStatusLabelValues...,
+		) // deprecated metric
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ch <- prometheus.MustNewConstMetric(
+			c.nodeDesc.nodeInfo,
+			prometheus.CounterValue,
+			1,
+			cpuArch,
+		)
+		for pkgID, val := range nodeEnergy.EnergyInCore.Stat {
+			// Node metrics in joules (counter)
+			ch <- prometheus.MustNewConstMetric(
+				c.nodeDesc.nodeCoreJoulesTotal,
+				prometheus.CounterValue,
+				float64(val.Aggr)/miliJouleToJoule,
+				pkgID, nodeName, "rapl",
+			)
+		}
+		for pkgID, val := range nodeEnergy.EnergyInUncore.Stat {
+			ch <- prometheus.MustNewConstMetric(
+				c.nodeDesc.nodeUncoreJoulesTotal,
+				prometheus.CounterValue,
+				float64(val.Aggr)/miliJouleToJoule,
+				pkgID, nodeName, "rapl",
+			)
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for pkgID, val := range nodeEnergy.EnergyInDRAM.Stat {
+			ch <- prometheus.MustNewConstMetric(
+				c.nodeDesc.nodeDramJoulesTotal,
+				prometheus.CounterValue,
+				float64(val.Aggr)/miliJouleToJoule,
+				pkgID, nodeName, "rapl",
+			)
+		}
+		for pkgID, val := range nodeEnergy.EnergyInPkg.Stat {
+			ch <- prometheus.MustNewConstMetric(
+				c.nodeDesc.nodePackageJoulesTotal,
+				prometheus.CounterValue,
+				float64(val.Aggr)/miliJouleToJoule,
+				pkgID, nodeName, "rapl",
+			)
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ch <- prometheus.MustNewConstMetric(
+			c.nodeDesc.nodePlatformJoulesTotal,
+			prometheus.CounterValue,
+			float64(nodeEnergy.SensorEnergy.Aggr())/miliJouleToJoule,
+			nodeName, "acpi",
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.nodeDesc.nodeOtherComponentsJoulesTotal,
+			prometheus.CounterValue,
+			float64(nodeEnergy.EnergyInOther.Aggr())/miliJouleToJoule,
+			nodeName,
+		)
+		if config.EnabledGPU {
+			for gpuID, val := range nodeEnergy.EnergyInGPU.Stat {
+				ch <- prometheus.MustNewConstMetric(
+					c.nodeDesc.nodeGPUJoulesTotal,
+					prometheus.CounterValue,
+					float64(val.Aggr)/miliJouleToJoule,
+					gpuID, nodeName, "nvidia",
+				)
+			}
+		}
+	}()
+}
+
+// updatePodMetrics send pod metrics to prometheus
+func (c *PrometheusCollector) UpdatePodMetrics(wg *sync.WaitGroup, ch chan<- prometheus.Metric) {
+	for _, pod := range podEnergy {
+		wg.Add(1)
```

Comment:
We are going to have pod*goroutines below added into wg?
any chance to make wg added by pod?

Reply:
We could create one gorotine per pod, or many per pod
Go is very efficient to handle a very large number of gorotines
So, I am creating multiple gorotines per pod to speedup the metric calculation process as much as possible

Reply:
https://github.com/sustainable-computing-io/kepler/pull/287/files#r990643209
as what I am worried about, so far with large number of gorotines, some var may calculated in many times as duplicated. I am not sure by one gorotine per pod we can avoid duplicate.

Reply:
Which variable will be calculated?
The goroutines will only read data...

Reply:
count `container.Command[:int(math.Min(float64(len(container.Command)), 10))`

Reply:
Good catch, I will update that!